### PR TITLE
Spec lazy-loading room members

### PR DESCRIPTION
**Pull request by turt2live**
**Reviewers**: Although this has team review requested, I would appreciate particular review from @ara4n as the proposal author and @KitsuneRal as having gone through the gauntlet. Feedback in addition is more than appreciated.

**Disclaimer**: Although the implementation is lacking in areas (see below), it is believed by the spec core team that lazy-loading has enough benefits to be added to the spec, as per the MSC guidelines.

----

Implementation references:
* https://github.com/matrix-org/synapse/pull/3574 - Room summary API impl
* https://github.com/matrix-org/synapse/pull/3568 - `/members` additions
* https://github.com/matrix-org/synapse/pull/3567 - `/context` becoming aware of LL
* https://github.com/matrix-org/synapse/pull/2970 - lazy-loading filters + `/sync` & `/messages` behaviour

Notable lacks of implementation:
* Lack of pagination token support on `/members` - https://github.com/matrix-org/synapse/blob/e26e6b3230f0b55376f0f3bf823dd789ac7064d0/synapse/rest/client/v1/room.py#L394
* Synapse always returns redundant members on filtered endpoints - https://github.com/matrix-org/synapse/issues/5393
* Synapse does not return left/banned users in `m.heroes` - https://github.com/matrix-org/synapse/issues/4926
* Synapse does not deduplicate redundant members when paginating `/messages` - https://github.com/matrix-org/synapse/blob/b54b03f9e1abc1964fe5f00115a165a2b8e10df5/synapse/handlers/pagination.py#L262
 * Synapse maintains its redundant membership event cache globally but should be per event stream - https://github.com/matrix-org/synapse/issues/5421

Spec references:
* https://github.com/matrix-org/matrix-doc/pull/1758 - Original spec
* Fixes https://github.com/matrix-org/matrix-doc/issues/1868
* Fixes https://github.com/matrix-org/matrix-doc/issues/1945
* [MSC1227](https://github.com/matrix-org/matrix-doc/issues/1227) - Lazy loading members
* [MSC688](https://github.com/matrix-org/matrix-doc/issues/688) - Room summaries (in a lazy loaded world)

Alterations from the proposals:
* The MSCs are unclear of what scope "redundant" is defined under. Written as "across any of the supported endpoints".
* Clarification that servers and clients should be careful when dealing with historical data. Unsure how Synapse/Riot handle this, if at all.
* Unclear if `summary` should be required. The implementation only does so when lazy-loading is active, however this seems like a bug. Written as required so that clients can use `m.heroes` to name the room.

(*note*: due to these proposals being older Google Doc MSCs, they have had the alterations described here represented in the applicable threaded comments instead of the body of the proposal. This is to hopefully more clearly represent the changes.)

### CODE REVIEW

File: api/client-server/message_pagination.yaml [hunk 1]
```diff
@@ -108,6 +111,23 @@ paths:
                   type: object
                   title: RoomEvent
                   "$ref": "definitions/event-schemas/schema/core-event-schema/room_event.yaml"
+              state:
+                type: array
+                description: |-
+                  A list of state events relevant to showing the ``chunk``. For example, if
```
Comment (bwindels) [inline api/client-server/message_pagination.yaml]: Can we mention as well that the server should always include the membership event for the syncing event when LL is enabled? See https://github.com/matrix-org/synapse/pull/3916 & https://github.com/vector-im/riot-web/issues/7282

File: api/client-server/sync.yaml [hunk 2]
```diff
@@ -34,6 +34,10 @@ paths:
         Clients use this API when they first log in to get an initial snapshot
         of the state on the server, and then continue to call this API to get
         incremental deltas to the state, and to receive new messages.
+
+        *Note*: this endpoint supports lazy-loading. See `Filtering <#filtering>`_
+        for more information. Lazy-loading members is only supported on a ``StateFilter``
+        for this endpoint.
```
Comment (bwindels) [inline api/client-server/sync.yaml]: Can we mention as well that the server should always include the membership event for the syncing user when LL is enabled? See https://github.com/matrix-org/synapse/pull/3916 & https://github.com/vector-im/riot-web/issues/7282

File: api/client-server/event_context.yaml [hunk 3]
```diff
@@ -34,6 +34,9 @@ paths:
         This API returns a number of events that happened just before and
         after the specified event. This allows clients to get the context
         surrounding an event.
+
+        *Note*: this endpoint supports lazy-loading. See `Filtering <#filtering>`_
```
Comment (jryans) [inline api/client-server/event_context.yaml]: Maybe the following style here and elsewhere?

```suggestion
        *Note*: This endpoint supports lazy-loading. See `Filtering <#filtering>`_
```

File: specification/client_server_api.rst [hunk 4]
```diff
@@ -1268,6 +1268,39 @@ Filters can be created on the server and can be passed as as a parameter to APIs
...
+Lazy-loading room members
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Membership events often take significant resources for clients to track. In an
+effort to reduce the number of resources used, clients can enable "lazy-loading"
+for room members. By doing this, servers will only ever send membership events
+which are relevant to the client.
+
+In terms of filters, this means enabling ``lazy_load_members`` on a ``StateFilter``
+or ``RoomEventFilter``. When enabled, lazy-loading aware endpoints (see below)
```
Comment (jryans) [inline specification/client_server_api.rst]: Is there supposed to be a list of endpoint that we are pointing to with "see below"? I can't seem to find it, but maybe I am missing something...

File: api/client-server/event_context.yaml (hunk 3 above)
Comment (turt2live) [reply api/client-server/event_context.yaml]: I have unreasonably strong opinions about this, mostly based on aesthetics and personal preference

File: specification/client_server_api.rst (hunk 4 above)
Comment (turt2live) [reply specification/client_server_api.rst]: yes, there should be...

File: api/client-server/event_context.yaml (hunk 3 above)
Comment (jryans) [reply api/client-server/event_context.yaml]: Most usages of "Note:" that I can find in the spec follow that with a capital letter, which is personally what I'd expect, since "Note:" is not really part of the sentence that follows. Anyway, I would just suggest it would be nice to choose one approach or the other for grammatical consistency.

Comment (turt2live) [reply api/client-server/event_context.yaml]: fair, I'll fix it.

File: api/client-server/sync.yaml (hunk 2 above)
Comment (turt2live) [reply api/client-server/sync.yaml]: This opened up a can of worms I wasn't expecting, but am summarizing it here before committing it in the formal language of the spec:

https://github.com/matrix-org/synapse/pull/3916 literally *always* included the syncing user's event, however this was affected by https://github.com/matrix-org/synapse/pull/3840 and https://github.com/matrix-org/synapse/pull/3936 in that LL does not appear to apply to incremental syncs. https://github.com/vector-im/riot-web/issues/7211#issuecomment-419976346 has more information. 

Given the proposal makes no mention of this, and the conclusions around 7211 seem unfinished, I think I want @ara4n to comment on this. The three questions being:
1. When do we LL our own membership?
2. Are we considered redundant?
3. Should the spec apply LL to incremental syncs? (it currently does in this PR)

Comment (bwindels) [reply api/client-server/sync.yaml]: Wrt to your 3rd question: As it stands, synapse sends all member state events since `from` for gappy syncs, e.g. doesn't lazy load. IMHO this is reasonable, but if we don't want to spec it like this, then this MSC might miss the Matrix 1.0 window, as changing and testing this in riot and synapse would take some time.

Comment (bwindels) [reply api/client-server/sync.yaml]: > synapse sends all member state events since `from` for gappy syncs, e.g. doesn't lazy load for incremental syncs.

Actually, it *does* do lazy loading for non-gappy syncs: If a member speaks for the first time since you initially synced, the server should send the member state event.

Comment (bwindels) [reply api/client-server/sync.yaml]: Did some digging in the synapse code to try and unblock this:

Wrt to question 1: the own membership is [included](https://github.com/matrix-org/synapse/blob/master/synapse/handlers/sync.py#L732) when the full state of the room would otherwise be returned, which is at least when doing an initial ([1](https://github.com/matrix-org/synapse/blob/master/synapse/handlers/sync.py#L1298),[2](https://github.com/matrix-org/synapse/blob/master/synapse/handlers/sync.py#L1624)) sync, and also when [joining a room](https://github.com/matrix-org/synapse/blob/master/synapse/handlers/sync.py#L1679).

Wrt to question 2: when the own membership is included in the response, it is [added to the cache](https://github.com/matrix-org/synapse/blob/master/synapse/handlers/sync.py#L858) of members that have been sent, which is kept per userid & deviceid, and cleared when doing an initial sync again. So yes, synapse considers our own membership redundant once sent down.

Comment (turt2live) [reply api/client-server/sync.yaml]: Thank you for looking into it - I am curious if that's the intention of the proposal as well. We have enough of lazy-loading implemented that I think a few lacking parts of the implementation is okay. The point of the implementation is to prove that the idea works in practice, and in this case I believe it has done so.

I do also agree that sending all membership events for gappy (`limited`) syncs makes sense - helps kinda flush the client's cache, although could be problematic in large rooms.

Comment (bwindels) [reply api/client-server/sync.yaml]: > I do also agree that sending all membership events for gappy (`limited`) syncs makes sense - helps kinda flush the client's cache, although could be problematic in large rooms.

It should only send the member events since the `from` parameter of the sync request to the start of the returned timeline, not *all* member events.

File: specification/client_server_api.rst [hunk 5]
```diff
@@ -1268,6 +1268,44 @@ Filters can be created on the server and can be passed as as a parameter to APIs
 which return events. These filters alter the data returned from those APIs.
 Not all APIs accept filters.
 
+Lazy-loading room members
```
Comment (richvdh) [inline specification/client_server_api.rst]: it's a shame that the general /filter API endpoints now appear under this heading. Suggest moving it to after the API? or giving the endpoints their own `~`-level heading?

File: api/client-server/event_context.yaml [hunk 6]
```diff
@@ -34,6 +34,9 @@ paths:
         This API returns a number of events that happened just before and
         after the specified event. This allows clients to get the context
         surrounding an event.
+
+        *Note*: This endpoint supports lazy-loading. See `Filtering <#filtering>`_
```
Comment (richvdh) [inline api/client-server/event_context.yaml]:
```suggestion
        *Note*: This endpoint supports lazy-loading of room member events. See `Filtering <#lazy-loading-room-members>`_
```

Comment (richvdh) [reply api/client-server/event_context.yaml]: Non-blocking thought: I feel like we would benefit from more information about how lazy-loading is applied to this endpoint. In particular: what does `include_redundant_members` mean here? However, we can always add that later.

File: api/client-server/message_pagination.yaml [hunk 7]
```diff
@@ -33,6 +33,9 @@ paths:
       description: |-
         This API returns a list of message and state events for a room. It uses
         pagination query parameters to paginate history in the room.
+
+        *Note*: This endpoint supports lazy-loading. See `Filtering <#filtering>`_
```
Comment (richvdh) [inline api/client-server/message_pagination.yaml]:
```suggestion
        *Note*: This endpoint supports lazy-loading of room member events. See `Filtering <#lazy-loading-room-members>`_
```

File: api/client-server/rooms.yaml [hunk 8]
```diff
@@ -288,6 +288,44 @@ paths:
...
+            each room by the sync API. Defaults to the current state of the room,
+            as determined by the server.
+          x-example: "YWxsCgpOb25lLDM1ODcwOA"
+        # XXX: As mentioned in MSC1227, replacing `[not_]membership` with a JSON
+        # filter might be a better alternative.
+        # See https://github.com/matrix-org/matrix-doc/issues/1337
+        - in: query
+          name: membership
+          type: string
+          enum:
```
Comment (richvdh) [inline api/client-server/rooms.yaml]: do we not need some alternative swagger incantation to say that you can set more than one of these? (assuming you can?)

File: api/client-server/sync.yaml [hunk 9]
```diff
@@ -34,6 +34,16 @@ paths:
         Clients use this API when they first log in to get an initial snapshot
         of the state on the server, and then continue to call this API to get
         incremental deltas to the state, and to receive new messages.
+
+        *Note*: This endpoint supports lazy-loading. See `Filtering <#filtering>`_
+        for more information. Lazy-loading members is only supported on a ``StateFilter``
+        for this endpoint. When lazy-loading is enabled, servers MUST include the
+        syncing user's own membership event when they join a room, or when the
+        full state of rooms is requested. The user's own membership event is eligible
+        for being considered redudant by the server. When a sync is ``limited``,
```
Comment (richvdh) [inline api/client-server/sync.yaml]: redudant

File: api/client-server/sync.yaml [hunk 10]
```diff
@@ -34,6 +34,16 @@ paths:
...
+        the server MUST return membership events for the timeline, even if the
```
Comment (richvdh) [inline api/client-server/sync.yaml]: what does it mean for an event to be in the timeline but not in the response? this is referring to events in the "gap" due to the limit? might be good to say that.

File: api/client-server/sync.yaml [hunk 11]
```diff
@@ -34,6 +34,16 @@ paths:
...
+        applicable events are not in the response, regardless as to whether or not
+        they are redundant.
```
Comment (richvdh) [inline api/client-server/sync.yaml]: does that mean that `include_redundant_members` is ignored for gappy syncs?

File: api/client-server/sync.yaml [hunk 12]
```diff
@@ -125,6 +137,50 @@ paths:
...
                       type: object
                       properties:
+                        summary:
+                          title: RoomSummary
+                          type: object
+                          description: |-
+                            Information about the room which clients may need to
+                            correctly render it to users.
+                          properties:
+                            "m.heroes":
```
Comment (richvdh) [inline api/client-server/sync.yaml]: are these properties not `required` ?

File: specification/client_server_api.rst [hunk 13]
```diff
@@ -1268,6 +1268,44 @@ Filters can be created on the server and can be passed as as a parameter to APIs
...
+which are relevant to the client.
+
+In terms of filters, this means enabling ``lazy_load_members`` on a ``StateFilter``
+or ``RoomEventFilter``. When enabled, lazy-loading aware endpoints (see below)
+will only include membership events for the ``sender`` of events being included
+in the response. For example, if a client makes a ``/sync`` request with lazy-loading
+enabled, the server will only return membership events for the ``sender`` of events
+in the timeline, not all members of a room.
+
+Repeated calls to lazy-loading aware endpoints will result in redundant membership
```
Comment (richvdh) [inline specification/client_server_api.rst]: This could really do with an example of wtf qualifies as a "redundant" membership event. I'm really struggling to wrap my head around it.

File: api/client-server/definitions/room_event_filter.yaml [hunk 14]
```diff
@@ -16,6 +16,19 @@ allOf:
...
+    lazy_load_members:
+      type: boolean
+      description: |-
+        If ``true``, enables lazy-loading of membership events. See
+        `Lazy-loading room members <#lazy-loading-room-members>`_
+        for more information. Defaults to ``false``.
+    include_redundant_members:
+      type: boolean
+      description: |-
+        If ``true``, enables redudant membership events. Does not
```
Comment (richvdh) [inline api/client-server/definitions/room_event_filter.yaml]: redudant

File: specification/client_server_api.rst [hunk 15]
```diff
@@ -1268,6 +1268,44 @@ Filters can be created on the server and can be passed as as a parameter to APIs
...
+Lazy-loading room members
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Membership events often take significant resources for clients to track. In an
+effort to reduce the number of resources used, clients can enable "lazy-loading"
+for room members. By doing this, servers will only ever send membership events
+which are relevant to the client.
+
+In terms of filters, this means enabling ``lazy_load_members`` on a ``StateFilter``
+or ``RoomEventFilter``. When enabled, lazy-loading aware endpoints (see below)
```
Comment (richvdh) [inline specification/client_server_api.rst]: I'm a bit confused as to what it means to enable ``lazy_load_members`` on a RoomEventFilter rather than a StateFilter :/

File: specification/modules/instant_messaging.rst [hunk 16]
```diff
@@ -278,70 +278,42 @@ choose a name:
...
+   on the members of the room. Clients should consider `m.room.member`_ events
+   for users other than the logged-in user, as defined below.
+
+   i. If the ``m.heroes`` for the room are greater or equal to
+      ``m.joined_member_count + m.invited_member_count - 1``, then use the
+      membership events for the heroes to calculate display names for the
+      users (`disambiguating them if required`_) and concatenating them. For
+      example, the client may choose to show "Alice, Bob, and Charlie
+      (@charlie:example.org)" as the room name. The client may optionally
+      limit the number
```
Comment (richvdh) [inline specification/modules/instant_messaging.rst]: the number of... ?

File: specification/client_server_api.rst (hunk 5 above)
Comment (turt2live) [reply specification/client_server_api.rst]: Will give them their own heading - that was an accident.

File: api/client-server/rooms.yaml (hunk 8 above)
Comment (turt2live) [reply api/client-server/rooms.yaml]: Implementation/MSC only supports one at a time, sadly. We do have a swagger extension to allow multiple though for when it comes up: `name: membership...`

File: api/client-server/sync.yaml (hunk 11 above)
Comment (turt2live) [reply api/client-server/sync.yaml]: yes, I think...

File: specification/client_server_api.rst (hunk 15 above)
Comment (turt2live) [reply specification/client_server_api.rst]: Only `/sync` supports it on a StateFilter. Every other lazy-loadable endpoint uses the RoomEventFilter. I guess I should have put those words into the spec...

File: specification/modules/instant_messaging.rst (hunk 16 above)
Comment (turt2live) [reply specification/modules/instant_messaging.rst]: I have no excuse.

File: api/client-server/rooms.yaml (hunk 8 above)
Comment (KitsuneRal) [reply api/client-server/rooms.yaml]: On the other hand, isn't specifying 2 entries here equivalent to specifying 1 entry in `not_membership`, and vice versa?

File: api/client-server/event_context.yaml (hunk 6 above)
Comment (turt2live) [reply api/client-server/event_context.yaml]: I think specification of `include_redundant_members` is handled sufficiently in the general section for lazy-loading, although it might need a few more words

File: api/client-server/rooms.yaml (hunk 8 above)
Comment (turt2live) [reply api/client-server/rooms.yaml]: There's 4 membership states (join, leave, ban, invite), so no matter what you're stuck with 1 being included/excluded against your preference. [MSC1337](https://github.com/matrix-org/matrix-doc/issues/1337) will need updating to fix this.

File: api/client-server/sync.yaml (hunk 12 above)
Comment (turt2live) [reply api/client-server/sync.yaml]: no, because if they are omitted then the client should interpret it as "no change".

File: specification/client_server_api.rst (hunk 13 above)
Comment (turt2live) [reply specification/client_server_api.rst]: I'm having troubles trying to clarify this, as it's hard to wrap my head around the algorithm. I suspect this needs someone who knows what they are doing to clarify it.

Comment (richvdh) [reply specification/client_server_api.rst]: I've had a look at the impl for this, but the confusion is not helped by #5395.

https://docs.google.com/document/d/11yn-mAkYll10RJpN0mkYEVqraTbU3U4eQx9MNrzqX1U/edit#heading=h.q5mz9uiufb2g says of it:

> By default the lazy_load_members setting deduplicates redundant members across the same series of /syncs.  However, if the client does want to receive each and every member on every /sync, it can turn include_redundant_members to true. 

That kinda makes sense: if we've just sent a client a member event for Bob, we don't need to send the same event in the next /sync just because he happened to say something again. 

I guess we can extend that to /messages too: we only need to include member events for senders that we have not recently sent in a sync. But we need to be clear about what counts as 'recently sent': for instance, if we backpaginate to a point where Bob changes avatar (so there is a new membership event), do we then resend his latest membership over /sync next time he says something? (I'm assuming the answer is an emphatic 'no').

File: api/client-server/message_pagination.yaml [hunk 17]
```diff
@@ -108,6 +111,23 @@ paths:
...
+                description: |-
+                  A list of state events relevant to showing the ``chunk``. For example, if
+                  lazy-loading members is enabled in the filter then this will contain any
+                  applicable membership events. Servers should be careful to not exclude
+                  membership events which are older than ones already sent to the client.
+                  Likewise, clients should be cautious and avoid using older membership
+                  events as the current membership event when paginating backwards.
+
+                  Unless ``include_redundant_members`` is ``true``, the server should remove
+                  redundant members which would have already been sent to clients in prior calls
```
Comment (richvdh) [inline api/client-server/message_pagination.yaml]:
```suggestion
                  membership events which would have already been sent to clients in prior calls
```

File: api/client-server/message_pagination.yaml [hunk 18]
```diff
@@ -108,6 +111,23 @@ paths:
                   type: object
                   title: RoomEvent
                   "$ref": "definitions/event-schemas/schema/core-event-schema/room_event.yaml"
+              state:
+                type: array
+                description: |-
+                  A list of state events relevant to showing the ``chunk``. For example, if
+                  lazy-loading members is enabled in the filter then this will contain any
+                  applicable membership events. Servers should be careful to not exclude
```
Comment (richvdh) [inline api/client-server/message_pagination.yaml]:
```suggestion
                  the membership events for the the senders of events in the ``chunk``. Servers should be careful to not exclude
```

File: api/client-server/message_pagination.yaml [hunk 19]
```diff
@@ -108,6 +111,23 @@ paths:
                   type: object
                   title: RoomEvent
                   "$ref": "definitions/event-schemas/schema/core-event-schema/room_event.yaml"
+              state:
+                type: array
+                description: |-
+                  A list of state events relevant to showing the ``chunk``. For example, if
+                  lazy-loading members is enabled in the filter then this will contain any
```
Comment (richvdh) [inline api/client-server/message_pagination.yaml]:
```suggestion
                  ``lazy_load_members`` is enabled in the filter then this will contain any
```

or "lazy-loading of membership events"

File: specification/client_server_api.rst (hunk 13 above)
Comment (richvdh) [reply specification/client_server_api.rst]: >  (I'm assuming the answer is an emphatic 'no').

though that is in conflict with https://github.com/matrix-org/matrix-doc/pull/2035/files#diff-addc31d0732ed525604677b40a2466c2R126 :/

File: api/client-server/sync.yaml [hunk 20]
```diff
@@ -34,6 +34,17 @@ paths:
...
         of the state on the server, and then continue to call this API to get
         incremental deltas to the state, and to receive new messages.
+
+        *Note*: This endpoint supports lazy-loading. See `Filtering <#filtering>`_
+        for more information. Lazy-loading members is only supported on a ``StateFilter``
+        for this endpoint. When lazy-loading is enabled, servers MUST include the
+        syncing user's own membership event when they join a room, or when the
+        full state of rooms is requested. The user's own membership event is eligible
+        for being considered redundant by the server. When a sync is ``limited``,
+        the server MUST return membership events for events in the gap (from ``since``),
```
Comment (richvdh) [inline api/client-server/sync.yaml]:
```suggestion
        the server MUST return membership events for events in the gap (between ``since`` and the start of the returned timeline),
```

File: api/client-server/sync.yaml [hunk 21]
```diff
@@ -34,6 +34,17 @@ paths:
...
+        even if the applicable events are not in the response, regardless as to whether
```
Comment (richvdh) [inline api/client-server/sync.yaml]:
```suggestion
        regardless as to whether
```

they are inherently not in the response if they are in the gap

File: api/client-server/sync.yaml [hunk 22]
```diff
@@ -34,6 +34,17 @@ paths:
...
+        or not they are redundant. ``include_redundant_members`` is ignored for limited
```
Comment (richvdh) [inline api/client-server/sync.yaml]: sorry, having reconsidered this, I don't think it is correct that ``include_redundant_members`` is (or should be) ignored. 


Imagine that we have a room where Bob is speaking. We sync, and get messages from Bob, so get his membership event in the state section. Then we go offline for a while, so there is going to be a gap, and come back later. If Bob spoke within the last few messages, do we need to resend his membership event? I don't think we do.

Indeed I don't see why we should necessarily include redundant membership events for events in the gap.

File: specification/client_server_api.rst [hunk 23]
```diff
@@ -1268,6 +1268,47 @@ Filters can be created on the server and can be passed as as a parameter to APIs
...
+to true in the filter.
+
+Servers should be cautious about which events they consider redundant. Membership
+events can change over time, and should be included as relevant to maintain the
+historical record. Likewise, clients should be cautious about treating an older event
+as the current membership event for a user.
+
+.. Note::
+   Repeated calls using the same filter to *any* lazy-loading aware endpoint may
+   result in redundant members being excluded from future calls. For example, a
```
Comment (richvdh) [inline specification/client_server_api.rst]:
```suggestion
   result in redundant membership events being excluded from future calls. For example, a
```

File: specification/client_server_api.rst [hunk 24]
```diff
@@ -1268,6 +1268,47 @@ Filters can be created on the server and can be passed as as a parameter to APIs
...
+   request to ``/sync`` followed by a request to ``/messages`` may result in a
+   future call to ``/sync`` excluding members included by the ``/messages`` call.
```
Comment (richvdh) [inline specification/client_server_api.rst]:
```suggestion
   future call to ``/sync`` excluding membership events returned by the ``/messages`` call.
```

File: specification/client_server_api.rst [hunk 25]
```diff
@@ -1268,6 +1268,47 @@ Filters can be created on the server and can be passed as as a parameter to APIs
...
+(or a ``StateFilter`` in the case of ``/sync`` only). When enabled, lazy-loading
+aware endpoints (see below) will only include membership events for the ``sender``
+of events being included in the response. For example, if a client makes a ``/sync``
+request with lazy-loading enabled, the server will only return membership events
+for the ``sender`` of events in the timeline, not all members of a room.
+
+Repeated calls to lazy-loading aware endpoints will result in redundant membership
+events being excluded by default. Clients often track which membership events they
+already have, therefore making the extra information not as useful to the client.
+Clients can always request redundant members by setting ``include_redundant_members``
```
Comment (richvdh) [inline specification/client_server_api.rst]:
```suggestion
Clients can always request redundant membership events by setting ``include_redundant_members``
```

File: api/client-server/sync.yaml (hunk 22 above)
Comment (turt2live) [reply api/client-server/sync.yaml]: I'm not entirely sure tbh - it seems like it's a method of trying to refresh the client's cache, however if the event hasn't changed then do we really need to send it again?

File: specification/client_server_api.rst (hunk 13 above)
Comment (turt2live) [reply specification/client_server_api.rst]: This is one of those points that needs a Matthew to comment on: do we de-duplicate across endpoints? what about for the case you've highlighted (backfill an avatar change)? I don't really have answers :(

File: api/client-server/definitions/room_event_filter.yaml [hunk 26]
```diff
@@ -16,6 +16,19 @@ allOf:
...
+    lazy_load_members:
+      type: boolean
+      description: |-
+        If ``true``, enables lazy-loading of membership events. See
+        `Lazy-loading room members <#lazy-loading-room-members>`_
+        for more information. Defaults to ``false``.
+    include_redundant_members:
+      type: boolean
+      description: |-
+        If ``true``, enables redundant membership events. Does not
```
Comment (ara4n) [inline api/client-server/definitions/room_event_filter.yaml]: Perhaps:
s/enables redundant membership events/sends all membership events for all events, even if they have recently been sent/
to avoid it being a complete tautology?

File: api/client-server/message_pagination.yaml [hunk 27]
```diff
@@ -108,6 +111,23 @@ paths:
                   type: object
                   title: RoomEvent
                   "$ref": "definitions/event-schemas/schema/core-event-schema/room_event.yaml"
+              state:
+                type: array
+                description: |-
+                  A list of state events relevant to showing the ``chunk``. For example, if
+                  ``lazy_load_members`` is enabled in the filter then this will contain any
+                  the membership events for the the senders of events in the ``chunk``. Servers should be careful to not exclude
```
Comment (ara4n) [inline api/client-server/message_pagination.yaml]: wordwrap? O:-)

File: api/client-server/message_pagination.yaml [hunk 28]
```diff
@@ -108,6 +111,23 @@ paths:
...
+                  membership events which are older than ones already sent to the client.
```
Comment (ara4n) [inline api/client-server/message_pagination.yaml]: hm, for what definition of 'old'? i'm failing to grok the rationale here or know how to interpret this.

File: api/client-server/message_pagination.yaml [hunk 29]
```diff
@@ -108,6 +111,23 @@ paths:
...
+                  Likewise, clients should be cautious and avoid using older membership
+                  events as the current membership event when paginating backwards.
+
+                  Unless ``include_redundant_members`` is ``true``, the server should remove
```
Comment (ara4n) [inline api/client-server/message_pagination.yaml]: i wonder if 'should' could be 'may' here - to highlight that it's perfectly valid for servers to deliberately send some redundant state  if it helps their implementation (i.e. to avoid having to try to comprehensively track what data the client has seen and what it hasn't)

File: api/client-server/sync.yaml (hunk 22 above)
Comment (ara4n) [reply api/client-server/sync.yaml]: the point of `include_redundant_members` is that it was added (i think at @turt2live's request?) purely as a mechanism for stupid clients which simply track no info about what members are currently in a room.  so i would expect that `include_redundant_members` is true, you get all the members sent regardless of anything else. 

So in a `limited` sync, you absolutely should send *all* the membership events which happened during the gap (in order not to miss people joining and then not speaking, for instance).  However, if `include_redundant_members` is false, you can then skip the membership events for the events the actual sync response which you know you've already sent to the client.  Whereas if it is true, you need to send them all because you consider the client amnesiac.

Tl:DR, we should just remove the `include_redundant_members` sentence here, i think.

File: api/client-server/sync.yaml [hunk 30]
```diff
@@ -125,6 +138,50 @@ paths:
...
+                            Information about the room which clients may need to
+                            correctly render it to users.
+                          properties:
+                            "m.heroes":
+                              type: array
+                              description: |-
+                                The users which can be used to generate a room name
+                                if the room does not have one. Required if the room
+                                does not have a ``m.room.name`` or ``m.room.canonical_alias``
+                                state event with non-empty content.
```
Comment (ara4n) [inline api/client-server/sync.yaml]: this double-negative is a bit of a headfrak.  How about:

> Required if the room has a missing or empty `m.room.name` and `m.room.canonical_alias` state event`

File: specification/client_server_api.rst (hunk 13 above)
Comment (ara4n) [reply specification/client_server_api.rst]: i'm kinda regretting agreeing to let `include_redundant_members` be a thing. as far as I know, nobody's actually used it, and it just complicates the spec. 

For normal LL (i.e. `include_redundant_members` being false),
 * When backpaginating /messages, we include a state block for the members present in that block of events (as of the beginning of the block of events).
 * In theory we can avoid sending membership events to the client which we know are redundant - i.e. ones where the membership hasn't changed for a given user since we last sent a membership event to the client in that /messages sequence.  Just like /sync, this is optional, and implementations may choose not to track precisely which memberships a given user has seen.  As noted at https://github.com/matrix-org/synapse/blob/b54b03f9e1abc1964fe5f00115a165a2b8e10df5/synapse/handlers/pagination.py#L262, this optimisation hasn't been implemented in Synapse yet (and i imagine @turt2live will want to note this on this PR's description comment).  As a result, /messages in LL mode uses slightly more bandwidth currently on Synapse because it includes all members for the events in the response in the state block of that response, even if it's redundant.
 * The reason we could get away without the state block prior to LL was because we already knew all the members in the room at the point we started back-paginating.

Now, if `include_redundant_members` is true, the state block must include the membership events for all the events present in that block whatever.  In other words, Synapse's current behaviour for /messages is the same as if `include_redundant_members` is true.

Hope that clears it up?

File: specification/modules/instant_messaging.rst [hunk 31]
```diff
@@ -278,70 +278,42 @@ choose a name:
...
+#. If neither of the above conditions are met, the client can optionally guess
+   an alias from the ``m.room.alias`` events in the room. This is a temporary
+   measure while clients do not promote canonical aliases as prominently. This
+   step may be removed in a future version of the specification.
 
+#. If none of the above conditions are met, a name should be composed based
+   on the members of the room. Clients should consider `m.room.member`_ events
+   for users other than the logged-in user, as defined below.
+
+   i. If the ``m.heroes`` for the room are greater or equal to
```
Comment (ara4n) [inline specification/modules/instant_messaging.rst]: If the **number of** `m.heroes`

File: specification/client_server_api.rst (hunk 13 above)
Comment (turt2live) [reply specification/client_server_api.rst]: > i'm kinda regretting agreeing to let include_redundant_members be a thing. as far as I know, nobody's actually used it, and it just complicates the spec.

I haven't really used it because up until writing this PR I didn't really understand how lazy-loading actually affects things. I do plan on using it though, as it's useful for bots which need to populate pills/mentions with profile metadata (and other things they might do).

>and i imagine @turt2live will want to note this on this PR's description comment

This is mentioned under "Notable lacks of implementation" already.

----

That clears it up for me though, thank you!

**Comment by turt2live** [comment]
Shoving this back into the queue to get a checkmark (hopefully) from someone on the team. There's a few people who have reviewed this, so picking on someone isn't super obvious.

**Comment by ara4n** [comment]
i'm happy with it from my pov fwiw, but i guess my rewrite needs some review if someone can face it.

**Comment by turt2live** [comment]
arguably this is fine to merge as I reviewed your edits, but I think a third opinion would be good given the amount of review on here.

**Comment by ara4n** [comment]
let's just merge it.
